crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -344,21 +344,30 @@ def upgrade_controller_nodes
         return upgrade_drbd_clusters unless drbd_nodes.empty?
 
         founder = ::Node.find("pacemaker_founder:true").first
-        cluster_env = founder[:pacemaker][:config][:environment]
+        cluster = founder[:pacemaker][:config][:environment]
 
-        non_founder = ::Node.find(
+        non_founder_nodes = ::Node.find(
           "pacemaker_founder:false AND " \
-          "pacemaker_config_environment:#{cluster_env}"
-        ).first
+          "pacemaker_config_environment:#{cluster}"
+        )
+        non_founder_nodes.select! { |n| !n.upgraded? }
```

Comment:
you are changing the `non_founder_nodes` set with this call. I'd prefer the node sets are immutable. In practice that would mean to have a new variable for `non_upgraded_nodes`. It will work the same way, but semantically I find it makes more sense.